src/commands/registerCommands.ts

```diff
@@ -179,24 +181,86 @@ async function grammarAssociationCommand(documentURI: Uri, languageClient: Langu
   const fileUri = await window.showOpenDialog(options);
   if (fileUri && fileUri[0]) {
     // The XSD, DTD has been selected, get the proper syntax for binding this grammar file in the XML document.
-    const identifier = TextDocumentIdentifier.create(documentURI.toString());
     const grammarURI = fileUri[0];
     try {
-      const result = await commands.executeCommand(ServerCommandConstants.ASSOCIATE_GRAMMAR_INSERT, identifier, grammarURI.toString(), bindingType);
-      // Insert the proper syntax for binding
-      const lspTextDocumentEdit = <TextDocumentEdit>result;
-      const workEdits = new WorkspaceEdit();
-      for (const edit of lspTextDocumentEdit.edits) {
-        workEdits.replace(documentURI, languageClient.protocol2CodeConverter.asRange(edit.range), edit.newText);
+      const currentFile = getCurrentFile();
+      if (bindingType == 'fileAssociation') {
+        // Bind grammar using file association
+        await bindWithFileAssociation(documentURI, grammarURI, currentFile);
+      } else {
+        // Bind grammar using standard binding
+        await bindWithStandard(documentURI, grammarURI, bindingType, languageClient);
       }
-      workspace.applyEdit(workEdits); // apply the edits
-
     } catch (error) {
       window.showErrorMessage('Error during grammar binding: ' + error.message);
     };
   }
 }
 
+/**
+ * Perform grammar binding using file association through settings.json
+ *
+ * @param documentURI the URI of the current XML document
+ * @param grammarURI the URI of the user selected grammar file
+ * @param currentFile the opened TextDocument
+ */
+async function bindWithFileAssociation(documentURI: Uri, grammarURI: Uri, currentFile: TextDocument) {
+  const absoluteGrammarFilePath = grammarURI.toString();
+  const currentFilename = getCurrentFileName(documentURI.toString());
+  const currentWorkspaceUri = getAbsoluteWorkspaceUri(currentFile);
+  const relativeGrammarFilePath = getRelativeUri(currentWorkspaceUri, absoluteGrammarFilePath);
+  const grammarFilePath = absoluteGrammarFilePath.includes(currentWorkspaceUri) ? relativeGrammarFilePath : absoluteGrammarFilePath;
```

Comment:
If I look at a directory path and a file path, where the file has the same name as the directory, but is outside the directory, e.g:
a/b/c/foo -> directory
a/b/c/foo.xml -> file, outside of above directory
In this case, the includes will be true and it will use the relative path when it shouldn't.
The solution here is to only compare the directories instead of the uri with the file